The Saved Objects import feature improves information transfer for multiple data sources (#2572)
opensearch-project/project-website: _posts/2024-02-08-enhancement-multiple-data-source-import-saved-object.md

---
layout: post
title:  "Launch highlight: The Saved Objects import feature improves information transfer for multiple data sources"
authors:
 - nluyu
 - amznyja

> date: 2024-02-08

Reviewer: Update publish date to tomorrow, 2/22/24

categories:
 - feature

> ---

Reviewer: Please update with the following meta: 
meta_keywords:  OpenSearch data source information, OpenSearch saved objects, OpenSearch multiple data sources, OpenSearch Dashboards

meta_description: Unlock seamless dashboard migration with OpenSearch. Its enhanced data source information ensures accurate visualizations across clusters and eliminates transfer errors between dashboards.

---
layout: post

> title:  "Launch Highlight: The Saved Objects import feature improves multiple data sources information transfer"

Reviewer 2: ```suggestion
title:  "Launch highlight: The Saved Objects import feature improves information transfer for multiple data sources"
```

authors:
 - nluyu
 - amznyja
date: 2024-02-08
categories:
 - feature
---

> OpenSearch Dashboards launched multiple data sources functionality in OpenSearch 2.4, empowering Dashboards to retrieve data from various compatible OpenSearch endpoints. With this feature enabled, you gain the ability to connect multiple data sources while adding sample data or querying data through the OpenSearch Dashboards Dev Tools console. However, a limitation arises when attempting to import saved objects by uploading a local file. This blog describes how the OpenSearch Project has solved that challenge.

Reviewer 2: ```suggestion
OpenSearch Dashboards launched support for multiple data sources in OpenSearch 2.4, empowering Dashboards to retrieve data from various compatible OpenSearch endpoints. With this feature enabled, you gain the ability to connect multiple data sources while adding sample data or querying data through the OpenSearch Dashboards Dev Tools console. However, a limitation arises when attempting to import saved objects by uploading a local file. This blog post describes how the OpenSearch Project has solved that challenge.
```

> When `data_source.enabled` is `true`, multiple data sources is active and exported objects include data source information. Consequently, when visualizing these objects in a dashboard, the visualization tool seamlessly recognizes their origins. Conversely, importing saved objects without enabling multiple data sources or from dashboards lacking data source information results in imported objects missing crucial data source details. This omission poses challenges for accurate visualization within dashboards using the multiple data sources functionality.  

Reviewer 2: ```suggestion
When `data_source.enabled` is `true`, the multiple data sources functionality is active and exported objects include data source information. Consequently, when visualizing these objects on a dashboard, the visualization tool seamlessly recognizes their origin. Conversely, importing saved objects without enabling multiple data sources or from dashboards lacking data source information results in imported objects missing crucial data source information. This omission poses challenges for accurate visualization within dashboards using the multiple data sources functionality.  
```

> To address this issue, you now have the option to specify a data source when importing saved objects. Objects without any data source information will have the specified dataSourceId attached to the `objectId`. In cases where saved objects contain different data source information, such as when importing from various sources, the `objectId` will already include the relevant data source information, ensuring consistency in behavior. For example, if connecting to dataSource1 and importing from dataSource2, the `objectId` already contains the relevant data source information. In this case, dataSourceId is replaced. Also, the ID schema remains consistent with how you use sample data and Dev Tools with multiple data sources enabled. 

Reviewer 2: ```suggestion
To address this issue, you now have the option to specify a data source when importing saved objects. Objects without any data source information will have the specified dataSourceId attached to the `objectId`. In cases where saved objects contain different data source information, such as when importing from various sources, the `objectId` will already include the relevant data source information, ensuring consistent behavior. For example, if connecting to dataSource1 and importing from dataSource2, the `objectId` already contains the relevant data source information. In this case, dataSourceId is replaced. Also, the ID schema remains consistent with how you use sample data and Dev Tools with multiple data sources enabled. 
```

## Benefits
Benefits of this enhancement include: 

> We want to provide user the ability to specify a data source when importing saved objects. When user import saved object without any data source information, the imported saved objects will attach the specified dataSourceId with the o`bjectId`; while user import saved objects different data source information, for example, user can connect with dataSource1 and some exported saved objects from dataSource2), then the `objectId` would already contains data source info ( dataSource2 ), for this case, we will replace the dataSourceId. The id schema are exactly same as when user add sample data and Dev Tools with multiple data source enabled. The behavior will keep consistent for users.

Reviewer 2: ```suggestion
```

> We want to provide user the ability to specify a data source when importing saved objects. When user import saved object without any data source information, the imported saved objects will attach the specified dataSourceId with the o`bjectId`; while user import saved objects different data source information, for example, user can connect with dataSource1 and some exported saved objects from dataSource2), then the `objectId` would already contains data source info ( dataSource2 ), for this case, we will replace the dataSourceId. The id schema are exactly same as when user add sample data and Dev Tools with multiple data source enabled. The behavior will keep consistent for users.

Reviewer 2: It appears that this paragraph is the original of the rewritten version above.

Author: If this will confuse readers, I think I can remove it

> * Providing imported saved objects with proper data source information, enhancing data visualization accuracy.

Reviewer 2: ```suggestion
* Providing imported saved objects with proper data source information, thus enhancing data visualization accuracy.
```

* Facilitating data migration by enabling the transfer of saved objects between dashboards, even across different clusters.

> ## Configuring and using saved objects importing

Reviewer 2: ```suggestion
## Configuring and importing saved objects
```

> ## Configuring and using saved objects importing

Reviewer 2: Confirm that my change retains the intended technical meaning.

Author: there is a tic mark in the middle of a word
